app/src/org/commcare/entity/AndroidAsyncNodeEntityFactory.kt

```diff
@@ -0,0 +1,116 @@
+package org.commcare.entity
+
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeout
+import org.commcare.CommCareApplication
+import org.commcare.cases.entity.AsyncNodeEntityFactory
+import org.commcare.cases.entity.Entity
+import org.commcare.cases.entity.EntityStorageCache
+import org.commcare.suite.model.Detail
+import org.commcare.suite.model.EntityDatum
+import org.commcare.tasks.PrimeEntityCacheHelper
+import org.commcare.util.LogTypes
+import org.javarosa.core.model.condition.EvaluationContext
+import org.javarosa.core.model.instance.TreeReference
+import org.javarosa.core.services.Logger
+import java.lang.RuntimeException
+
+/**
+ * Android Specific Implementation of AsyncNodeEntityFactory
+ * Uses [PrimeEntityCacheHelper] to prime entity cache blocking the user when required
+ */
+class AndroidAsyncNodeEntityFactory(
+    d: Detail,
+    private val entityDatum: EntityDatum?,
+    ec: EvaluationContext?,
+    entityStorageCache: EntityStorageCache?
+) : AsyncNodeEntityFactory(d, ec, entityStorageCache) {
+
+    companion object {
+        const val TEN_MINUTES = 10 * 60 * 1000L
+    }
+
+    override fun prepareEntitiesInternal(
+        entities: MutableList<Entity<TreeReference>>
+    ) {
+        if (detail.shouldOptimize() && detail.isCacheEnabled) {
```

Comment:
There was another place where a similar check was done, which implied there were 3 potential states for this system based on combinations of different configs. I'm a little concerned it'll be hard to the various places these modes are checked in sync over time, it might be helpful to unify the checks in one place so that if changes are made later we can't miss one.

Reply:
https://github.com/dimagi/commcare-android/pull/2928/commits/71981614565917191d1fe077982d9633cbc31411